incubator/hnc/controllers/hierarchy_controller.go

```diff
@@ -203,20 +209,26 @@ func (r *HierarchyReconciler) updateObjects(ctx context.Context, log logr.Logger
 func (r *HierarchyReconciler) syncWithForest(ctx context.Context, log logr.Logger, inst *tenancy.Hierarchy) []string {
 	r.Forest.Lock()
 	defer r.Forest.Unlock()
-
-	ns := r.Forest.AddOrGet(inst.ObjectMeta.Namespace)
 	affected := []string{}
 	conds := []tenancy.Condition{}
 
-	// Sync our data structures with the current parent. The current parent might not exist (if, for
-	// example, the hierarchy is being created as a result of `kubectl apply -f` on a directory); in
-	// this case, just set a condition on the child, which will be removed once the parent exists.
-	//
-	// TODO: detect when a missing parent gets added so we can remove this condition.
-	curParent := r.Forest.Get(inst.Spec.Parent)
-	if inst.Spec.Parent != "" && curParent == nil {
-		log.Info("Couldn't set parent", "reason", "missing", "parent", inst.Spec.Parent)
-		conds = append(conds, tenancy.Condition{Msg: "missing parent"})
+	// Get the in-memory namespace and mark it as existing. If this is the first time we're
+	// reconciling this namespace, mark its relatives as being affected since they may have been
+	// waiting for this parent.
+	ns := r.Forest.Get(inst.ObjectMeta.Namespace)
+	if ns.Create() {
+		log.Info("Found new namespace")
+		affected = append(affected, ns.RelativesNames()...)
+	}
+
+	// Sync this namespace with its current parent.
+	var curParent *forest.Namespace
```

Comment:
I am a little confused about ns deletion workflow.
If a ns is the parent of child namespaces, what will happen if this parent ns is deleted? Who is responsible to change the parent field of the singleton in all child namespaces to nil? or nothing will happen besides child ns always stays in missing parent condition.

Reply:
Right now, the child will simply stay in the "missing parent" condition. We may add the ability to add an `ownersReference` from the child to the parent, which would automatically delete the child when the parent is deleted as well. However, by default, if the parent is deleted, the child just sticks around.